### ISSUE TITLE
[train] Running ray cluster sometimes get 0 metrics when training dataset

### ISSUE THREAD

**Issue by Alxe1** [ray-project/ray#28061]
### What happened + What you expected to happen

I run my code with ray cluster successfully,  but it somtimes get 0 result using auc metrics:
```
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4513) Epoch 19/20
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4514) Epoch 19/20
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4512) Epoch 19/20
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4516) Epoch 19/20
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4515) Epoch 19/20
70/70 [==============================] - 4s 52ms/step - loss: 0.0580 - auc: 0.9979
(BaseWorkerMixin pid=4511) Epoch 19/20
70/70 [==============================] - 4s 51ms/step - loss: 0.0533 - auc: 0.9981
(BaseWorkerMixin pid=4513) Epoch 20/20
 2/70 [..............................] - ETA: 1s - loss: 0.0114 - auc: 0.0000e+00
(BaseWorkerMixin pid=4514) Epoch 20/20
 2/70 [..............................] - ETA: 1s - loss: 0.0114 - auc: 0.0000e+00
(BaseWorkerMixin pid=4512) Epoch 20/20
 1/70 [..............................] - ETA: 0s - loss: 0.0119 - auc: 0.0000e+00
(BaseWorkerMixin pid=4516) Epoch 20/20
 1/70 [..............................] - ETA: 0s - loss: 0.0119 - auc: 0.0000e+00
(BaseWorkerMixin pid=4515) Epoch 20/20
 2/70 [..............................] - ETA: 1s - loss: 0.0114 - auc: 0.0000e+00
(BaseWorkerMixin pid=4511) Epoch 20/20
70/70 [==============================] - 3s 50ms/step - loss: 0.0088 - auc: 0.0000e+00
result: [{'loss': [1.342253565788269, 0.12474628537893295, 0.014094097539782524, 0.6787583827972412, 0.3139781355857849, 0.3230475187301636, 0.3027200698852539, 0.0612882599234581, 0.15591417253017426, 0.09106193482875824, 0.11201471835374832, 0.10636495798826218, 0.010080278851091862, 0.00976957380771637, 0.19445857405662537, 0.06490223854780197, 0.16993646323680878, 0.057986412197351456, 0.05330931022763252, 0.008758490905165672], 'auc': [0.0, 0.0, 0.0, 0.6334323883056641, 0.944059431552887, 0.920643150806427, 0.8857136964797974, 0.9939171671867371, 0.9281724095344543, 0.9952266216278076, 0.9929127097129822, 0.982867419719696, 0.0, 0.0, 0.9533206224441528, 0.9964863657951355, 0.9621405005455017, 0.997946560382843, 0.9981012940406799, 0.0]}
```

I don't understand why it gets that result, and how can I deal with it. 

### Versions / Dependencies

ray: 1.13.0
pyarrow: 9.0.0
tensorflow: 2.3.0
python: 3.7.10

### Reproduction script

```

def train_func(config):
    os.environ["HDF5_USE_FILE_LOCKING"] = "FALSE"  
    per_work_batch_size = config.get("batch_size", 64)
    epochs = config.get("epochs", 3)
    steps_per_epoch = config.get("steps_per_epoch", 70)

    tf_config = json.loads(os.environ["TF_CONFIG"])
    num_works = len(tf_config["cluster"]["worker"])

    strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()

    global_batch_size = per_work_batch_size * num_works

    with strategy.scope():
        multi_worker_model = build_and_compile_cnn_model(config)

    dataset_shard = ray.train.get_dataset_shard()

    tf_dataset = to_df2(
        dataset_shard,
        output_types=(tf.float32, tf.int32),
        output_shapes=(tf.TensorShape((None, config.get("sparse_num"))), tf.TensorShape((None,))),
        label_column="is_click",
        batch_size=global_batch_size,
        drop_last=True
    )
    # AutoShardPolicy OFF
    tf_dataset = prepare_dataset_shard(tf_dataset)

    history = multi_worker_model.fit(
        tf_dataset,
        epochs=epochs,
        steps_per_epoch=steps_per_epoch
    )
    multi_worker_model.save("/multi_worker_model.h5", save_format="h5")

    result = history.history
    return result


def train_tf_mnist(spark_base, num_workers=2, use_gpu=False, epochs=3):
    parquet_path = "s3://my-bucket/tmp_train"
    # user_num, item_num, sparse_num, cate_num = write_spark_df_to_parquet(spark_base, parquet_path)

    dataset = ray.data.read_parquet(parquet_path, parallelism=8)
    print(dataset.take(10))

    config = {"user_num": 33356, "item_num": 5736, "user_item_dim": 32, "sparse_num": 407,
              "cate_num": 61, "feature_embed_dim": 16, "embed_norm": 0.001, "hidden_units": [64, 32, 32],
              "dropout_ratio": 0.2, "cross_num": 3, "batch_size": 128, "epochs": epochs}

    trainer = Trainer(backend="tensorflow", num_workers=num_workers, use_gpu=use_gpu)
    trainer.start()
    result = trainer.run(
        train_func=train_func,
        config=config,
        dataset=dataset
    )

    trainer.shutdown()
    print(f"result: {result}")


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("--address", required=False, type=str, help="The address to use for ray")
    parser.add_argument("--num_workers", "-n", type=int, default=6, help="Sets number of workers for training")
    parser.add_argument("--use_gpu", action="store_true", default=False, help="Enable GPU training")
    parser.add_argument("--epochs", type=int, default=20, help="Number of epochs to train for")
    parser.add_argument("--smoke-test", action="store_true", default=False, help="Finish quickly for testing")

    args, _ = parser.parse_known_args()
    print("======================================args smoke_test: {}".format(args.smoke_test))
    print("======================================args.num_workers: {}".format(args.num_workers))
    print("======================================args.epochs: {}".format(args.epochs))

    import ray

    spark_base = SparkBase()

    if args.smoke_test:
        ray.init(num_cpus=2)
        train_tf_mnist(spark_base)
    else:
        ray.init(address=args.address)
        train_tf_mnist(
            spark_base,
            num_workers=args.num_workers,
            use_gpu=args.use_gpu,
            epochs=args.epochs
        )

    model = load_model("/multi_worker_model.h5", custom_objects={"CrossLayer": CrossLayer})


```

**Comment by xwjiang2010** [comment]
Smells like some type casting issues. Can you make sure that `to_df2` is generating tf tensor correctly? Is the output label having the correct type?

**Comment by Alxe1** [comment]
> Smells like some type casting issues. Can you make sure that `to_df2` is generating tf tensor correctly? Is the output label having the correct type?

Yes, and I test it, the convert code is in follow:
```
def to_df2(ray_dataset, output_types, output_shapes, label_column=None, batch_size=5, drop_last=False):
    """tensorflow2.3.0"""
    def make_generator():
        for batch in ray_dataset.iter_batches(batch_size=batch_size, batch_format="pandas", drop_last=drop_last):
            if label_column:
                y = batch.pop(label_column)
                x = np.vstack(list(batch.values.flatten()))
                yield x, y
            else:
                x = np.vstack(list(batch.values.flatten()))
                yield x

    dataset = tf.data.Dataset.from_generator(make_generator, output_types=output_types, output_shapes=output_shapes)

    return dataset
```

**Comment by matthewdeng** [comment]
@Alxe1 hey just checking are you still running into this issue?

**Comment by Alxe1** [comment]
> @Alxe1 hey just checking are you still running into this issue?

Now I'm using Ray2.0, it's not happened again.

**Comment by matthewdeng** [comment]
Great to hear.

Based on the values over time I feel like it may just be due to randomness of the model weights...

```
'auc': [0.0, 0.0, 0.0, 0.6334323883056641, 0.944059431552887, 0.920643150806427, 0.8857136964797974, 0.9939171671867371, 0.9281724095344543, 0.9952266216278076, 0.9929127097129822, 0.982867419719696, 0.0, 0.0, 0.9533206224441528, 0.9964863657951355, 0.9621405005455017, 0.997946560382843, 0.9981012940406799, 0.0]}
```

In any case, I'll close this issue for now but please re-open if you run into this again!
